Sources/Network/NetworkAuthenticator.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  Authenticator.swift
+//  Alicerce
+//
+//  Created by Luís Portela on 03/07/2017.
+//  Copyright © 2017 Mindera. All rights reserved.
+//
+
+import Foundation
+
+public protocol NetworkAuthenticator {
+    func authenticate(request: URLRequest, _ completion: @escaping (URLRequest) -> Void)
+
+    func shouldRetry(with data: Data?, response: HTTPURLResponse?, error: Swift.Error?) -> Bool
```

Comment:
Don't know if it makes sense, but this closure parameters are somehow "tied" to `URLSession` completion parameters. So it made me think if this protocol could have an `associatedtype` with the completion "type", that would be used for the retry closure.

What do you say? 😇